pkg/openstack/instances.go

```diff
@@ -707,5 +753,47 @@ func getAttachedPorts(client *gophercloud.ServiceClient, serverID string) ([]por
 		DeviceOwner: "compute:nova",
 	}
 
-	return openstack.GetPorts(client, listOpts)
+	type portWithTrunkDetails struct {
+		ports.Port
+		trunk_details.TrunkDetailsExt
+	}
+
+	var interfaces []ports.Port
+	var allPorts []portWithTrunkDetails
+
+	allPages, err := ports.List(client, listOpts).AllPages()
+	if err != nil {
+		return interfaces, err
+	}
+	err = ports.ExtractPortsInto(allPages, &allPorts)
+	if err != nil {
+		return interfaces, err
+	}
+
+	for _, portExt := range allPorts {
+		interfaces = append(interfaces, portExt.Port)
+		// Check if trunk is attached to the port
+		if portExt.TrunkDetails.TrunkID == "" {
+			continue
+		}
+		// Get subport addresses
+		for _, subport := range portExt.TrunkDetails.SubPorts {
+			p, err := ports.Get(client, subport.PortID).Extract()
+			if err != nil {
+				klog.Errorf("Failed to get subport %s details: %v", subport.PortID, err)
+				continue
+			}
+			n, err := networks.Get(client, p.NetworkID).Extract()
+			if err != nil {
+				klog.Errorf("Failed to get subport %s network details: %v", subport.PortID, err)
+				continue
+			}
+			p.Status = "ACTIVE"
+			p.NetworkID = n.Name
+			interfaces = append(interfaces, *p)
+		}
+	}
+
+	klog.V(5).Infof("Node %s has %d interfaces '%v'", serverID, len(interfaces), interfaces)
+	return interfaces, nil
```

Comment:
Looking at how they're used above, I wonder if we should return the return signature of this method to:

```go
func getAttachedPorts(client *gophercloud.ServiceClient, serverID string) (attachedPorts []ports.Port, subPorts []ports.Ports, err error)
```

And we return direct attached and subports separately. This would mean we wouldn't need the extremely confusing and error prone length conditional above. We could rename `addExtraNodeAddresses` to `addSubPortNodeAddresses` and call it unconditionally with the returned subports. If they're empty or nil it just wouldn't do anything.

Reply:
Hmm, since the caller of getAttachedPorts does not care the additional level of details I am not sure changing its signature is needed (right thing to do), I can for sure change addExtraNodeAddresses to addSubPortNodeAddresses, for code clarity.

Reply:
The critical difference is that we subsequently treat subports differently to directly attached ports because the former are not included in server addresses but the latter are. By returning them as a single un-differentiated slice we're making our lives more complicated later when we have to determine which is which.

Reply:
An alternative might be to not fetch subports here at all and just return directly attached ports. We could scan for subports later when we unconditionally call `addSubPortNodeAddresses`. That would require changing the return signature of this function to return `[]portWithTrunkDetails`.

Reply:
The alternative sounds reasonable, will check and update.

Reply:
The alternative seems a better idea, and it results in the same code structure as in the older releases, PR is updated.